### PR TITLE
Add postprocess plugin set

### PR DESCRIPTION
**Pull request by alexlafroscia**
Adds a new option to provide a set of PostCSS plugins to run after concatenation has taken place. This is useful for plugins like `postcss-sprites`, which should be run later in the CSS pipeline.

Closes salsify/broccoli-css-modules#3

## Todo

- [x] ~~Unit tests for the new Broccoli plugin~~ replaced by a dependency
- [x] Rebase from `master`
- [x] Smoke test to ensure the `postprocess` plugins are applied to the output file.

### CODE REVIEW

**Comment by alexlafroscia** [comment]
Still need to add tests, though:

- [ ] Unit tests for the new Broccoli plugin
- [ ] Smoke test to ensure the `postprocess` plugins are applied to the output file.

**Comment by dfreeman** [comment]
I probably won't have a chance to look at this deeply until this weekend, but thank you so much for taking this on!

**Comment by alexlafroscia** [comment]
No problem 😄  we're just using my fork in our app right now and I can confirm it works as intended. I'm excited to get this finished and published, we've been using a fork of `ember-css-modules` for a long time and I'm stoked to be able to just track the latest from here on out.

File: lib/output-styles-preprocessor.js [hunk 1]
```diff
@@ -24,8 +25,17 @@ OutputStylesPreprocessor.prototype.toTree = function(inputNode, inputPath, outpu
 
   debug('concatenating module stylesheets: %o', concatOptions);
 
+  var parentOptions = options.registry.app.options || {};
+  var cssModulesOptions = parentOptions.cssModules || {};
+  var postprocessPlugins = cssModulesOptions.plugins && cssModulesOptions.plugins.postprocess;
```
Comment (dfreeman) [inline lib/output-styles-preprocessor.js]: You can call [`this.owner.getPlugins()`](https://github.com/alexlafroscia/ember-css-modules/blob/master/index.js#L87-L89) to simplify this

File: lib/postcss-persistent-filter.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+PostCSS.prototype.targetExtension = 'css';
+
+PostCSS.prototype.processString = function processString(content, relativeFile) {
+  return this._processor.process(content, {
+    from: relativeFile,
+    to: relativeFile,
+    map: {
+      inline: false,
+      annotation: false
+    }
```
Comment (dfreeman) [inline lib/postcss-persistent-filter.js]: Will this play nicely with user-specified source map settings?

File: package.json [hunk 3]
```diff
@@ -24,6 +24,7 @@
   },
   "homepage": "https://github.com/salsify/ember-css-modules#readme",
   "devDependencies": {
+    "broccoli-persistent-filter": "^1.2.11",
```
Comment (dfreeman) [inline package.json]: This should be in `dependencies`, right?

File: lib/postcss-persistent-filter.js (hunk 2 above)
Comment (alexlafroscia) [inline lib/postcss-persistent-filter.js]: Not sure -- @jasonmit ?

File: lib/output-styles-preprocessor.js (hunk 1 above)
Comment (alexlafroscia) [inline lib/output-styles-preprocessor.js]: Oh awesome, I'll update that.

File: package.json (hunk 3 above)
Comment (alexlafroscia) [inline package.json]: Hmm, probably. @jasonmit ?

**Comment by alexlafroscia** [comment]
Sorry this got away from me @dfreeman. If you know of a good way to do the smoke test for the `postprocess` set, setting that up would be extremely helpful. I'll handle the unit test for the broccoli plugin tonight to make sure that it works right.

File: lib/postcss-persistent-filter.js (hunk 2 above)
Comment (jasonmit) [inline lib/postcss-persistent-filter.js]: Unsure what is meant by user specific map settings, these: https://ember-cli.com/asset-compilation#source-maps ?

File: package.json (hunk 3 above)
Comment (jasonmit) [inline package.json]: Yes, should be moved

File: lib/postcss-persistent-filter.js (hunk 2 above)
Comment (dfreeman) [inline lib/postcss-persistent-filter.js]: Yep, we [read that config](https://github.com/salsify/ember-css-modules/blob/master/index.js#L119-L126) and [pass it along](https://github.com/salsify/ember-css-modules/blob/master/lib/modules-preprocessor.js#L40) to `broccoli-css-modules` (as well as using it [when concatenating everything together](https://github.com/salsify/ember-css-modules/blob/master/lib/output-styles-preprocessor.js#L119))

**Comment by FransTwisk** [comment]
I hope this will be merged eventually. Very useful.

**Comment by alexlafroscia** [comment]
Yeah, sorry about that. getting back to the tests just hasn't been a priority but should be. It won't be this weekend, but maybe the weekend after I'll be able to carve out some time

**Comment by FransTwisk** [comment]
Any updates on this?

**Comment by alexlafroscia** [comment]
Yeah, so... No. But I really need to do this, since our app requires this PR and we'd really like to get on the official release (and PostCSS 6).  My holdup is mostly just not really knowing how to test Broccoli stuff. Maybe @jasonmit has some insight there?

**Comment by dfreeman** [comment]
For testing Broccoli stuff I've found [broccoli-fixture](https://github.com/broccolijs/broccoli-fixture) to be super helpful — we use it over in `broccoli-css-modules`, though that project is still on Mocha instead of QUnit for its tests.

**Comment by alexlafroscia** [comment]
Cool, I'll definitely check into that. First step is rebasing, since it's been so long since this PR was made 😅  but it's on my list to check this out this weekend since we really want to get our app off of PostCSS 5 and onto 6.

**Comment by alexlafroscia** [comment]
@dfreeman after checking out the new Plugin system you have added since this PR started (which is awesome, by the way!) I _think_ this could actually be added as a plugin instead. Would you prefer that to adding it to the default add-on? I feel like the use-cases for this are actually pretty slim, and like the idea of allowing people to opt in to running the code by installing an additional add-on.

File: lib/output-styles-preprocessor.js (hunk 1 above)
Comment (alexlafroscia) [inline lib/output-styles-preprocessor.js]: @dfreeman was this API removed between the initial PR and now? That method doesn't exist anymore

**Comment by dfreeman** [comment]
@alexlafroscia I think framing this as a plugin is an interesting idea — as it sounds like you've discovered, the current plugin API is lacking the hook you'd need, but I could easily imagine adding broccoli-based plugin methods for manipulating the tree pre-CSS-modules, post-CSS-modules and post-concatenation.

Your plugin would just implement that third hook, taking the concatenated file and performing an additional round of PostCSS processing on it. What do you think of an approach like that?

File: lib/output-styles-preprocessor.js (hunk 1 above)
Comment (dfreeman) [inline lib/output-styles-preprocessor.js]: I think that's right—IIRC that was renamed to distinguish between ECM plugins and PostCSS plugins

**Comment by alexlafroscia** [comment]
Yeah, that sounds interesting! I'm fine doing it that way.

My first commit here (I squashed all the previous commits together to make it easier to rebase onto `master`) implements things basically the same way that the initial PR did. The only major difference is that instead of including the Broccoli plugin that will get run on post-process as part of this repo, I found a separate one that was built on the same `broccoli-persistent-filter` that was already tested, and used that instead.

**Comment by dfreeman** [comment]
Given how lightweight this wound up being, I'm not opposed to landing it as core functionality, as long as we can get some kind of smoke test in place in one of the dummy app/addons to prevent regressions.

Regardless, it does feel like plugins _should_ be able to implement something like this, but I can save tackling that for another day :)

**Comment by alexlafroscia** [comment]
Working on that test now :) Just trying to look through the existing tests to get an idea of how I can verify that the `postprocess` plugins are actually run.

**Comment by alexlafroscia** [comment]
@dfreeman what would you be looking for for a "smoke test" here? I'm not sure how to verify that those plugins were run other than setting up a dummy app with a plugin and checking to see if the output CSS was transformed. Would that do? Maybe using [`ember-cli-addon-tests`](https://github.com/tomdale/ember-cli-addon-tests)?

**Comment by dfreeman** [comment]
I think the simplest thing to do would be to add a component to the dummy app whose `styles.css` relies on something you configure as a postprocess plugin. Then you can just throw together an integration test to verify the right style actually gets applied, similar to how we [verify that imports resolve correctly](https://github.com/salsify/ember-css-modules/blob/master/tests/acceptance/imports-test.js#L24).

**Comment by alexlafroscia** [comment]
👍 great idea. I'll do that!

**Comment by alexlafroscia** [comment]
No idea if I did that right... I'm on a new laptop and was having trouble getting the tests running locally. I'm going to let CI tell me if that worke.

**Comment by alexlafroscia** [comment]
@dfreeman just wanted to bump this... we'd love to get off of my old fork and onto the latest version, but we need this to land first!

**Comment by dfreeman** [comment]
Oops, sorry for the delay @alexlafroscia! I meant to look at this before I left for vacation and totally lost track. I should have some time in the next day or so to take a look.

**Comment by alexlafroscia** [comment]
No problem! Thanks for merging, glad to get this landed after so long (my fault, not yours) 😅

**Comment by dfreeman** [comment]
This is now released in 0.7.3 — thanks again @alexlafroscia! 🎉

**Pull request merged**